app_flutter/lib/status_grid.dart

```diff
@@ -44,122 +45,35 @@ class StatusGridContainer extends StatelessWidget {
-  /// This ensures a column is allocated in [_createTaskMatrix].
-  ///
-  /// When a new [taskColumnKey] is found, it is inserted and the index is incremeted.
-  Map<String, int> _createTaskColumnKeyIndex(List<CommitStatus> statuses) {
-    final Map<String, int> taskColumnKeyIndex = <String, int>{};
-    int currentIndex = 0;
-
-    /// O(Tasks * CommitStatuses). In production, this is usually O(85 * 100) ~ 8500 operations.
-    for (CommitStatus status in statuses) {
-      for (Stage stage in status.stages) {
-        for (Task task in stage.tasks) {
-          final String key = _taskColumnKey(task);
-          if (taskColumnKeyIndex.containsKey(key)) {
-            continue;
-          }
-
-          taskColumnKeyIndex[key] = currentIndex++;
-        }
+  /// Return how many [Task] since the last failure for [Column].
+  /// 
+  /// If no failure has ever occurred, return the highest possible value for
+  /// the matrix. This max would be the number of rows in the matrix.
+  int _lastFailed(task_matrix.Column a) {
+    for (int row = 0; row < a.tasks.length; row++) {
+      if (a.tasks[row]?.status == TaskBox.statusFailed) {
+        return row;
       }
     }
 
-    return taskColumnKeyIndex;
-  }
-
-  /// Create a matrix of [Task] for easier sorting of [List<CommitStatus>].
-  ///
-  /// If no [Task] can be placed in a cell of the matrix, it will be left null.
-  List<List<Task>> _createTaskMatrix(
-      List<CommitStatus> statuses, Map<String, int> taskColumnKeyIndex) {
-    /// Rows are commits, columns are [Task] with same [taskColumnKey].
-    final List<List<Task>> taskMatrix = List<List<Task>>.generate(
-        statuses.length, (_) => List<Task>(taskColumnKeyIndex.keys.length));
-
-    for (int statusIndex = 0; statusIndex < statuses.length; statusIndex++) {
-      final CommitStatus status = statuses[statusIndex];
-      final List<Task> statusTasks = taskMatrix[statusIndex];
-
-      /// Organize [Task] in [CommitStatus] to the column they map to.
-      for (Stage stage in status.stages) {
-        for (Task task in stage.tasks) {
-          statusTasks[taskColumnKeyIndex[_taskColumnKey(task)]] = task;
-        }
-      }
-    }
-
-    return taskMatrix;
-  }
-
-  /// Create [List<Task>] for [List<TaskIcon>] at the top of [StatusGrid].
-  List<Task> _createTaskIconRow(
-      List<CommitStatus> statuses, Map<String, int> taskColumnKeyIndex) {
-    final List<Task> taskIconRow = List<Task>(taskColumnKeyIndex.keys.length);
-
-    for (int statusIndex = 0; statusIndex < statuses.length; statusIndex++) {
-      final CommitStatus status = statuses[statusIndex];
-      for (Stage stage in status.stages) {
-        for (Task task in stage.tasks) {
-          final int taskIndex = taskColumnKeyIndex[_taskColumnKey(task)];
-          if (taskIconRow[taskIndex] == null) {
-            taskIconRow[taskIndex] = task;
-          }
-        }
-      }
-    }
-
-    return taskIconRow;
+    return a.tasks.length;
```

Comment:
This method is called O(nlogn) times during the sorting and there's only O(n) columns,  so a small optimization is to save the answer without re-calculating.

Even better, I'd make Column be immutable after creation, and let `lastFailed` be a constant property of Column. The spirit is to encapsulate low level details (lastFailed) in the low level class (Column), and let high level class (StatusGrid) decided how to utilize it.

Reply:
With this performance improvement, I think we should hold off until the team has explored the v2 dashboard. This generalization makes it easy to easily implement [the different ordering techniques we discussed in the design doc](https://docs.google.com/document/d/1kwh48xYuoVtadYFb2I2t6BJMGEW07K7T5WYcieAZZiY/edit#heading=h.7a1ri2tf3g6a). If the team finds some ordering techniques to be much more useful than others, we can optimize for them by implementing these const properties as needed.

Do you have any resources I can look into for making an object immutable after it has been created?

Reply:
SG.

Sorry, not immutable, my idea was to make Column.tasks read only. That's not a big deal though.